### PR TITLE
Tab completion

### PR DESCRIPTION
**Pull request by Carpetsmoker**
If set to False, it will load completion on next/prev (ie. the Tab key) instead
of showing by default.

### CODE REVIEW

File: qutebrowser/completion/completionwidget.py [hunk 1]
```diff
@@ -255,3 +257,26 @@ def showEvent(self, e):
...
+        try:
+            cmd.cursorPositionChanged.disconnect(cmd.update_completion)
+            cmd.textChanged.disconnect(cmd.update_completion)
+        # Don't fail if not connected
+        except TypeError:
+            pass
+
+    def _open_completion_if_needed(self):
+        if config.get('completion', 'auto-open') == False:
+            update_completion = pyqtSignal()
```
Comment (The-Compiler) [inline qutebrowser/completion/completionwidget.py]: What's this supposed to do? I don't think there's any valid use of `pyqtSignal()` outside of a class body, and here that's actually a signal on `cmd`. Actually I'd be surprised if it doesn't raise an exception, but I haven't tried yet.

Comment (Carpetsmoker) [inline qutebrowser/completion/completionwidget.py]: This is just some cruft lying around; it doesn't so anything.

Note that in `mainwindow/statusbar/command.py` does the same thing; but I don't think that does anythere there, either.

Comment (The-Compiler) [inline qutebrowser/completion/completionwidget.py]: I don't understand your last sentence - you mean some similar thing is in `command.py`? I don't see any `pyqtSignal` outside of the class body there.

**Comment by Carpetsmoker** [comment]
Wait, it's assigning `self.update_completion`... I've been programming Ruby for over a year, so excuse the confusion O.o

File: qutebrowser/completion/completionwidget.py (hunk 1 above)
Comment (Carpetsmoker) [inline qutebrowser/completion/completionwidget.py]: Wait, it's assigning `self.update_completion`... I've been programming Ruby for over a year, so excuse the confusion O.o

Comment (Carpetsmoker) [inline qutebrowser/completion/completionwidget.py]: Wait, it's assigning self.update_completion... I've been programming Ruby for over a year, so excuse the confusion O.o

**Comment by The-Compiler** [comment]
I noticed two small issues when testing this:

* When setting `auto-open` to true again interactively, the signals should be connected again so the completion shows up - but they aren't. You could rename `set_enabled` to `on_config_changed`, adjust the `objreg.get('config').changed.connect(...)` call, and then handle this there.
* When the completion is shown because tab is pressed, but then is hidden again (e.g. because accidentally doing `:se `, it doesn't show up again after pressing backspace - I think it should, right?

**Comment by Carpetsmoker** [comment]
> When the completion is shown because tab is pressed, but then is hidden again (e.g. because accidentally doing :se, it doesn't show up again after pressing backspace - I think it should, right?

I don't quite follow this? If I press:

    :se<Tab><Backspace>

completion still works (as I would expect).

**Comment by The-Compiler** [comment]
Try `:se<Tab><Space><Backspace>`. I think that's what triggered it for me. The completion is hidden because there are no completions for `:se ` and then not show again.

**Comment by Carpetsmoker** [comment]
Thanks. Fixed both points.

**Comment by The-Compiler** [comment]
This seems to break completion somehow :wink:

I tried `:set comp<tab>` and then the completion just hangs there (with `auto-open` set to true).

File: qutebrowser/completion/completionwidget.py [hunk 2]
```diff
@@ -255,3 +260,21 @@ def showEvent(self, e):
         if scrollbar is not None:
             scrollbar.setValue(scrollbar.minimum())
         super().showEvent(e)
+
+    def _open_completion_if_needed(self):
+        self._set_auto_open(not config.get('completion', 'auto-open'))
```
Comment (The-Compiler) [inline qutebrowser/completion/completionwidget.py]: I don't really get what's happening here - does this actually *open* something like the completion says? Why the negation?

**Comment by Carpetsmoker** [comment]
Hm, the `update_completion` is scheduled with a timer and run *after* `self._next_prev_item`, which is not what we want (this took some time to figue out).

I did a basic fix; which seems to work. I'm not sure if this is the best way to fix it, though...

**Comment by The-Compiler** [comment]
From a quick look it looks good. I'll do some testing and merge this if everything is okay tomorrow (I hope).

The reason the timer is there is that it's a noticeable performance difference to postpone the update compared to doing it right away - when the user starts typing quickly, it won't immediately lag.

File: qutebrowser/completion/completer.py [hunk 3]
```diff
@@ -47,6 +47,7 @@ def __init__(self, cmd, win_id, parent=None):
         self._win_id = win_id
         self._cmd = cmd
         self._cmd.update_completion.connect(self.schedule_completion_update)
+        self._cmd.update_completion_now.connect(self.update_completion)
```
Comment (The-Compiler) [inline qutebrowser/completion/completer.py]: Note to self: I probably should update those signals after this is merged:

* `update_completion` -> `schedule_completion_update`
* `update_completion_now` -> `update_completion`

I guess that would've reduced your confusion?

File: qutebrowser/completion/completionwidget.py [hunk 4]
```diff
@@ -89,6 +89,7 @@ class CompletionView(QTreeView):
     # https://github.com/The-Compiler/qutebrowser/issues/117
 
     resize_completion = pyqtSignal()
+    connected = None
```
Comment (The-Compiler) [inline qutebrowser/completion/completionwidget.py]: This shouldn't be here but `self.connected = None` in `__init__` instead, otherwise that'd be an attribute which is set this way in *every* instance of that class (i.e. a class rather than an instance variable).

I'd also say let's make it "private" (i.e. `_connected`) and please add a quick note about what it is to the docstring of the class.

File: qutebrowser/completion/completionwidget.py [hunk 5]
```diff
@@ -255,3 +261,25 @@ def showEvent(self, e):
...
         super().showEvent(e)
+
+    def _open_completion_if_needed(self):
+        if not config.get('completion', 'auto-open'):
+            self._set_auto_open(True)
+
+    def _set_auto_open(self, enabled):
+        cmd = objreg.get('status-command', scope='window', window=self._win_id)
+
+        if enabled:
```
Comment (The-Compiler) [inline qutebrowser/completion/completionwidget.py]: If you press `<Tab>` multiple times (with `auto-open = False`), this will cause the signals to be connected again every time, thus calling `cmd.update_completion` multiple times every time there's a change.

Shouldn't the `connected` conditional below be here instead, so all this is only ran if it wasn't run before?

File: qutebrowser/completion/completionwidget.py [hunk 6]
```diff
@@ -255,3 +261,25 @@ def showEvent(self, e):
...
+            cmd.cursorPositionChanged.connect(cmd.update_completion)
+            cmd.textChanged.connect(cmd.update_completion)
+            if self.connected == None:
```
Comment (The-Compiler) [inline qutebrowser/completion/completionwidget.py]: Why `None` and not `False`?

**Comment by Carpetsmoker** [comment]
Those were exactly the sort of minor issues I wanted to look at, but sort of forgot :-( I fixed a typo as an apology.

File: qutebrowser/browser/network/qutescheme.py [hunk 7]
```diff
@@ -153,7 +153,7 @@ def qute_help(win_id, request):
             url=request.url().toDisplayString(),
             error="This most likely means the documentation was not generated "
                   "properly. If you are running qutebrowser from the git "
-                  "repository, please run scripts/asciidoc2html.py."
+                  "repository, please run scripts/asciidoc2html.py. "
```
Comment (The-Compiler) [inline qutebrowser/browser/network/qutescheme.py]: Seems like an unrelated change - is this actually fixing something or is it an accident?

**Comment by The-Compiler** [comment]
Hmm, I wonder if it really makes sense to have all this code in `completion.py` when it actually just does things related to `command.py`. I found this rather confusing when reading the code...

Maybe `_completion_connected`/`_set_auto_open`/`_open_completion_if_needed` should actually be in `command.py` and some signals added to `completer.py` which trigger those (e.g. an `open_completion_if_needed` signal for `completion_item_{prev,next}`).

File: qutebrowser/browser/network/qutescheme.py (hunk 7 above)
Comment (Carpetsmoker) [inline qutebrowser/browser/network/qutescheme.py]: Yes, it fixes a missing space ;-)

                  "repository, please run scripts/asciidoc2html.py."
                  "If you're running a released version this is a bug, please "

**Comment by The-Compiler** [comment]
I think it's time for a bit a bigger refactoring here, but I'll take care of that once v0.3 is out - see #760.

I'll leave some more comments here based on the current structure.

File: qutebrowser/browser/network/qutescheme.py (hunk 7 above)
Comment (The-Compiler) [inline qutebrowser/browser/network/qutescheme.py]: Indeed, thanks!

File: qutebrowser/completion/completer.py (hunk 3 above)
Comment (The-Compiler) [inline qutebrowser/completion/completer.py]: This is probably obsolete with #760.

File: qutebrowser/completion/completionwidget.py [hunk 8]
```diff
@@ -93,15 +95,16 @@ class CompletionView(QTreeView):
     def __init__(self, win_id, parent=None):
         super().__init__(parent)
         self._win_id = win_id
+        self._completion_connected = False
```
Comment (The-Compiler) [inline qutebrowser/completion/completionwidget.py]: This is initially `False`, but the signals *are* connected in `statusbar/command.py` if `auto-open` is true - wouldn't that double-connect the signals?

**Comment by The-Compiler** [comment]
Could you take a look at #765 maybe? This should do the same, but with the functionality mainly implemented in `completer.py` which is the better place IMHO.

**Comment by The-Compiler** [comment]
Closing this in favor of #765.
